File: 12-IntegerToRoman/main.go
package main

import "fmt"

func intToRoman(num int) string {
	type RomanPair struct {
		Value  int
		Symbol string
	}

	romanMap := []RomanPair{
		{1000, "M"},
		{900, "CM"},
		{500, "D"},
		{400, "CD"},
		{100, "C"},
		{90, "XC"},
		{50, "L"},
		{40, "XL"},
		{10, "X"},
		{9, "IX"},
		{5, "V"},
		{4, "IV"},
		{1, "I"},
	}

	result := ""
	for i := 0; i < len(romanMap); i++ {
		for num >= romanMap[i].Value {
			num -= romanMap[i].Value
			result += romanMap[i].Symbol
		}
	}

	return result
}

func main() {
	fmt.Println(intToRoman(3749))
	fmt.Println(intToRoman(58))
	fmt.Println(intToRoman(1994))
}

File: 12-IntegerToRoman/main.ts
const intToRoman = (num: number): string => {
  type RomanPair = {
    value: number;
    symbol: string;
  };

  const romanMap: RomanPair[] = [
    { value: 1000, symbol: "M" },
    { value: 900, symbol: "CM" },
    { value: 500, symbol: "D" },
    { value: 400, symbol: "CD" },
    { value: 100, symbol: "C" },
    { value: 90, symbol: "XC" },
    { value: 50, symbol: "L" },
    { value: 40, symbol: "XL" },
    { value: 10, symbol: "X" },
    { value: 9, symbol: "IX" },
    { value: 5, symbol: "V" },
    { value: 4, symbol: "IV" },
    { value: 1, symbol: "I" },
  ];

  let result: string = "";
  for (let i = 0; i < romanMap.length; i++) {
    while (num >= romanMap[i].value) {
      num -= romanMap[i].value;
      result += romanMap[i].symbol;
    }
  }

  return result 
};

console.log(intToRoman(3749));
console.log(intToRoman(58));
console.log(intToRoman(1994));
